src/js/reducers/historyReducer.ts

```diff
@@ -1,30 +1,36 @@
-/* eslint new-cap:0 */
-'use strict';
-
-var dl = require('datalib'),
-    ACTIONS = require('../actions/Names'),
-    LIMIT = 20,
-    IMPLICIT_BATCH = [ACTIONS.SET_SIGNAL, ACTIONS.UPDATE_GUIDE_PROPERTY],
-    BATCH_INTERVAL = 500,  // ms to identify new batch for same implicit action.
-    batch = 0,
-    prevAction, prevTime;
-
-function historyReducer(reducer) {
-  return function(state, action) {
+import {List} from 'immutable';
+import {AnyAction} from 'redux';
+import {getType} from 'typesafe-actions';
+import * as datasetActions from '../actions/datasetActions';
+import * as guideActions from '../actions/guideActions';
+import * as historyActions from '../actions/historyActions';
+import * as signalActions from '../actions/signalActions';
+import {VisState} from '../store';
+
+const dl = require('datalib');
+const LIMIT = 20;
+const IMPLICIT_BATCH = [getType(signalActions.setSignal), getType(guideActions.updateGuideProperty)];
+const BATCH_INTERVAL = 500;  // ms to identify new batch for same implicit action.
+let batch = 0;
+let prevAction;
+let prevTime;
+
+export function undoable(reducer) {
+  return function(state: VisState, action: AnyAction): VisState {
     if (typeof state === 'undefined') {
       state = create([], reducer(undefined, {}), []);
```

Comment:
I'm seeing some type inconsistencies in this reduce. `past`, `present`, and `future` should all be arrays (and thus use `.length`) or Immutable `List`s (and use `.size`). Looking through this code, it seems like we might be mixing between these two types. I'm fine with either choice of collection.

Reply:
great catch! this got fixed in a later commit, where i ended up going with immutable lists.
 https://github.com/vega/lyra/pull/427/commits/17a77abb1711b96c5c1e300c4d671c0d0e63cf1c#diff-2255b1a147d62cf636821d5ae360a49c